core/src/providers/google_vertex_ai.rs

```diff
@@ -0,0 +1,527 @@
+        assert!(n == 1);
+
+        if let Some(m) = max_tokens {
+            if m == -1 {
+                let tokens = self.encode(prompt).await?;
+                max_tokens = Some((self.context_size() - tokens.len()) as i32);
+            }
+        }
+
+        let api_key =
+            get_access_token(self.service_account_json.as_ref().unwrap().as_str()).await?;
+
+        let uri = format!(
+            "{}/publishers/google/models/{}:streamGenerateContent?alt=sse",
+            self.uri.clone().unwrap(),
+            self.id()
+        );
+
+        let c = streamed_chat_completion(
+            uri,
+            api_key,
+            &vec![Content {
+                role: String::from("USER"),
+                parts: vec![Part {
+                    text: String::from(prompt),
+                }],
+            }],
+            &vec![],
+            None,
+            temperature,
+            stop,
+            max_tokens,
+            match top_p {
+                Some(t) => t,
+                None => 1.0,
+            },
+            None,
+            event_sender,
+        )
+        .await?;
+
+        Ok(LLMGeneration {
+            created: utils::now(),
+            provider: ProviderID::GoogleVertexAi.to_string(),
+            model: self.id().clone(),
+            completions: vec![Tokens {
+                text: c.candidates[0].content.parts[0].text.clone(),
+                tokens: Some(vec![]),
+                logprobs: Some(vec![]),
+                top_logprobs: None,
+            }],
+            prompt: Tokens {
+                text: prompt.to_string(),
+                tokens: Some(vec![]),
+                logprobs: Some(vec![]),
+                top_logprobs: None,
+            },
+        })
+    }
+
+    async fn chat(
+        &self,
+        messages: &Vec<ChatMessage>,
+        _functions: &Vec<ChatFunction>,
+        _function_call: Option<String>,
+        temperature: f32,
+        top_p: Option<f32>,
+        n: usize,
+        stop: &Vec<String>,
+        mut max_tokens: Option<i32>,
+        _presence_penalty: Option<f32>,
+        _frequency_penalty: Option<f32>,
+        _extras: Option<Value>,
+        event_sender: Option<UnboundedSender<Value>>,
+    ) -> Result<LLMChatGeneration> {
+        assert!(self.service_account_json.is_some());
+        assert!(self.uri.is_some());
+        assert!(n == 1);
+
+        if let Some(m) = max_tokens {
+            if m == -1 {
+                max_tokens = None;
+            }
+        }
+
+        let api_key =
+            get_access_token(self.service_account_json.as_ref().unwrap().as_str()).await?;
+
+        let uri = format!(
+            "{}/publishers/google/models/{}:streamGenerateContent?alt=sse",
+            self.uri.clone().unwrap(),
+            self.id()
+        );
+
+        let c = streamed_chat_completion(
+            uri,
+            api_key,
+            &messages
+                .iter()
+                .map(|m| Content {
```

Comment:
I had to do something similar for Mistral, I used the `try_from` trait. It's probably a pattern that we can use here?

https://github.com/dust-tt/dust/blob/49c60579c74f393c0e09d37213008f45d3ac3e9f/core/src/providers/mistralai.rs#L174-L184

Reply:
TIL ! Will do